Redundancy backport

filipleple opened the pull request:
Backporting the entire upstream redundancy path train to our fork

So far targeting only VP66XX

EDK: https://github.com/Dasharo/edk2/pull/286
*ref: prot-1806*

philipanda committed:
ifdtool: Add set top swap size PCH strap subcommand

Top-Block Swap mode of Intel PCH allows to swap the boot block with
another location placed directly below it by redirecting the memory
accesses.

The range of the addresses to be redirected is configured using the Top
Swap Block Size (or BOOT_BLOCK_SIZE) PCH strap using 3 bits to encode
one of 8 sizes:
    64 KB, 128 KB, 256 KB, 512 KB, 1 MB, 2 MB, 4 MB or 8 MB.

The source and target ranges depend on the configured size, eg:
- 64 KB  - FFFF_0000h - FFFF_FFFFh -> FFFE_0000h - FFFE_FFFFh
- 128 KB - FFFE_0000h - FFFF_FFFFh -> FFFC_0000h - FFFD_FFFFh
- 8 MB   - FF80_0000h - FFFF_FFFFh -> FF00_0000h - FF7F_FFFFh

Only supporting Alder Lake-P and Alder Lake-N for now.

Needed for the bootblock redundancy feature suggested at
https://mail.coreboot.org/archives/list/coreboot@coreboot.org/thread/C6JN2PB7K7D67EG7OIKB6BBERZU5YV35/

TEST=check using xxd, MFIT tool, ensure VP6670 boots
Test details:
xxd:
  ./util/ifdtool/ifdtool -p adl -T 0x10000 vp66xx_fd.bin && \
  xxd vp66xx_fd.bin > vp66xx.hex && \
  xxd vp66xx_fd.bin.new > vp66xx_fd.new.hex && \
  diff -au vp66xx_fd.hex vp66xx_fd.new.hex

File vp66xx_fd.bin is 4096 bytes
Writing new image to vp66xx_fd.bin.new

Change-Id: I7f2ae67808e8ac0aaa2a9ad5e091c2b358d4a5ab

--- vp66xx_fd.hex       2025-10-08 12:03:09.527193533 +0200
+++ vp66xx_fd.new.hex   2025-10-08 12:05:08.717108142 +0200
@@ -18,7 +18,7 @@
 00000110: 7f78 0700 0000 0000 1800 0000 0000 1f00  .x..............
 00000120: 0808 1170 0000 0000 0000 7f06 80f8 8107  ...p............
 00000130: 0000 0000 0f00 0000 2222 2222 2202 2222  ........""""".""
-00000140: 0000 0000 0000 0000 0000 ff00 6000 80c8  ............`...
+00000140: 0000 0000 0000 0000 0000 ff00 0000 80c8  ................
 00000150: 4586 0036 0000 0000 0002 5800 0000 4000  E..6......X...@.
 00000160: 0018 0000 0000 0000 0000 0000 0000 0000  ................
 00000170: 0000 0000 0000 0000 54b3 04a0 3000 0140  ........T...0..@

mfittool:
./mfit --gui -decompose protectli_vp66xx_v0.9.2.rom
In the UI:
Flash Settings > BIOS Configuration > Top Swap Block Size
shows the value changing to the expected one, ie.
    -T 0x10000 results in 64kB
    -T 0x20000 results in 128kB
    -T 0x400000 results in 4MB
    etc.

Change-Id: I50e9d4160ee4b60e83567bcd33c9d80d428cf2bb
Upstream-Status: Backport [CB:89438]
Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>
Reviewed-on: https://review.coreboot.org/c/coreboot/+/89438
Reviewed-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>
Tested-by: build bot (Jenkins) <no-reply@coreboot.org>

philipanda committed:
intel soc,southbridge: Add Kconfig to set TSBS in IFD during build

To modify the Top Swap Block Size in the FD (if provided and
CONFIG_HAVE_IFD_BIN=y), set the following Kconfig variables:
- CONFIG_INTEL_HAS_TOP_SWAP
- CONFIG_INTEL_ADD_TOP_SWAP_BOOTBLOCK
- CONFIG_INTEL_TOP_SWAP_BOOTBLOCK_SIZE
- CONFIG_INTEL_IFD_SET_TOP_SWAP_BOOTBLOCK_SIZE

Needed for the bootblock redundancy feature suggested at
https://mail.coreboot.org/archives/list/coreboot@coreboot.org/thread/C6JN2PB7K7D67EG7OIKB6BBERZU5YV35/

TEST=build VP66xx with custom Kconfig, check if TSBS is modified in FD

Change-Id: I94d3d3e2511a7e56392a9e34f845ae91602ce7f1
Upstream-Status: Backport [CB:89493]
Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>
Reviewed-on: https://review.coreboot.org/c/coreboot/+/89493
Reviewed-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>
Tested-by: build bot (Jenkins) <no-reply@coreboot.org>

filipleple committed:
ifittool: allow adding files from a separate region

Allows ifittool to add FIT entries for files that are located in a
different FMAP region than the FIT table.

The region from where to source the file can be specified with -R.
If not given it defaults to using the value of the mandatory -r,
for full backwards compatibility.

Example: Tested with a custom binary with the bootblock and
corresponding FIT table in a separate region, and the microcode still
in the COREBOOT region:

λ ./ifittool -f test_ts.rom -a -n cpu_microcode_blob.bin -t 1 \
	-r BOOTBLOCK \
	-R COREBOOT \
	-s 4

Change-Id: I7e49247f280ec118e09cf173795d7602a4c0d7f6
Upstream-Status: Backport [CB:89608]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>
Reviewed-on: https://review.coreboot.org/c/coreboot/+/89608
Reviewed-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>
Tested-by: build bot (Jenkins) <no-reply@coreboot.org>

filipleple committed:
Makefile.mk: separate bootblocks into BOOTBLOCK and TOPSWAP

Add Kconfig INTEL_TOP_SWAP_SEPARATE_REGIONS. When enabled, place the
regular bootblock in BOOTBLOCK and the Top Swap bootblock in TOPSWAP
to simplify A B updates. This lays groundwork for redundancy where one
bootblock remains a read only golden copy and the other is replaceable.

No swap control logic is added in this change. The option depends on
INTEL_ADD_TOP_SWAP_BOOTBLOCK and defaults to n so existing builds are
unchanged. A custom .fmd is required with BOOTBLOCK and TOPSWAP added
at the end of the image.

Background and update flow are described here:
Link: https://mail.coreboot.org/archives/list/coreboot@coreboot.org/thread/C6JN2PB7K7D67EG7OIKB6BBERZU5Y
V35/

TEST=Build and run Protectli VP6650 (ADL-P), boots successfully with
correct microcode

Change-Id: I489406dd8d08ad85bb46324d3d009acb49b6c52a
Upstream-Status: Backport [CB:89570]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>
Reviewed-on: https://review.coreboot.org/c/coreboot/+/89570
Tested-by: build bot (Jenkins) <no-reply@coreboot.org>
Reviewed-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

filipleple committed:
src/mainboard/protectli/vault_adl_p/Kconfig: Capsule Update config

Enable capsule update with Top Swap redundancy

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: Ib183a1f72ee8585b2c4ad4376344de33ff54cbb9
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
src/mainboard/protectli/vault_adl_p/: Redundancy configuration

Configuration needed to enable the recently upstreamed Top Swap
redundancy feature

Change-Id: Ib183a1f72ee8585b2c4ad4376344de33ff54cbb9
Upstream-Status: Inappropriate [Dasharo downstream]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
src/southbridge/intel/common/firmware/Makefile.mk: fix after cherrypick

Fix an #endif mismatch after cherry-picking a gerrit patch

Change-Id: Iad2722b474b0919d60731dbe2df6e9b571927c30
Upstream-Status: Inappropriate [Dasharo downstream]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
soc/intel/common/block/rtc/rtc.c: control Top Swap via CMOS option

Toggle the RTC BUC control bit for Top Swap bootblock selection based on
the "attempt_slot_b" flag CMOS option, allowing to select which of the
BOOTBLOCK or TOP_SWAP regions to boot from.

This means that after an update, the CMOS option can be set to boot from
the newer TOP_SWAP bootblock. In case of failure, CMOS can be cleared to
revert to the known-good base BOOTBLOCK.

This is part of ongoing implementation of a redundancy feature proposed
on the mailing list:
https://mail.coreboot.org/archives/list/coreboot@coreboot.org/thread/C6JN2PB7K7D67EG7OIKB6BBERZU5YV35/

Switching between identical bootblocks doesn't impact further boot flow,
i.e. selecting which FMAP region to load consecutive stages from.
That is to be enabled in following patches.

So far tested and enabled for the Alder Lake SoC.

TEST=Boot VP6650, setting the attempt_slot_b flag to different values,
observing that it resets/continues booting correctly.

Change-Id: Ib183a1f72ee8585b2c4ad4376344de33ff54cbb9
Upstream-Status: Backport [CB:90042]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
soc/intel/common/block/rtc/rtc.c: Top Swap: add Slot B selection mechanism

If the Top Swap mechanism is enabled, after running the bootblock from
the TOP_SWAP region, boot from an updatable COREBOOT_TS FMAP region.

Having flashed the TOP_SWAP bootblock and COREBOOT_TS, this allows the
user to boot a newer version of the firmware with the ability to
revert to the previous known-good version by performing a CMOS reset.

Requires having a read-write COREBOOT_TS region in the FMAP file.

This is part of an ongoing implementation of a redundancy feature
proposed on the mailing list:
https://mail.coreboot.org/archives/list/coreboot@coreboot.org/thread/C6JN2PB7K7D67EG7OIKB6BBERZU5YV35/

TEST=Boot Protectli VP6650, setting the attempt_slot_b flag to
different values, observing the "Booting from COREBOOT/COREBOOT_TS
region" prints correspondingly.

Change-Id: Ieadc9bfbe940cbec79eb84f16a5d622bfbb82ede
Upstream-Status: Backport [CB:90147]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
security/lockdown/lockdown.c: option to lock COREBOOT and BOOTBLOCK

Add an option to lock the regions BOOTBLOCK and COREBOOT, leaving the
regions TOPSWAP and COREBOOT_TS for updates in an Intel Top Swap
redundancy scenario.

This means that the user can now write and choose to boot from the
update regions, selecting the attempt_slot_b CMOS option, and there is
a protected golden copy of the entire firmware, that cannot be
overwritten and can be reverted to by resetting CMOS.

This is part of an ongoing implementation of a redundancy feature
proposed on the mailing list:
https://mail.coreboot.org/archives/list/coreboot@coreboot.org/thread/C6JN2PB7K7D67EG7OIKB6BBERZU5YV35/

Change-Id: Ia6dea22c41e2fc778af6ca7049b72c92686ec85f
Upstream-Status: Backport [CB:90413]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>
Reviewed-on: https://review.coreboot.org/c/coreboot/+/90413
Tested-by: build bot (Jenkins) <no-reply@coreboot.org>
Reviewed-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
sb/intel/common/firmware/Makefile.mk: fix INTEL_IFD_SET_TOP_SWAP_BOOTBLOCK_SIZE

Halfway through the review of CB:89493 (commit 6e4501661095 ("intel
soc,southbridge: Add Kconfig to set TSBS in IFD during build")) the
option was renamed and lost "CONFIG_" prefix in the Makefile.  Add the
missing prefix.

The omission was discovered and the fix tested while trying to use this
option on Protectli VP6670 where incorrect Top Swap size prevented a
boot.

Adding a missing space while at it.

Change-Id: Ie85fc0b81b2231760878306cf065598bec390e9f
Upstream-Status: Backport [CB:90432]
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>
Reviewed-on: https://review.coreboot.org/c/coreboot/+/90432
Reviewed-by: Matt DeVillier <matt.devillier@gmail.com>
Tested-by: build bot (Jenkins) <no-reply@coreboot.org>

SergiiDmytruk committed:
src/Kconfig: add MAINBOARD_NEEDS_CMOS_OPTIONS

The addition provides a way to select USE_OPTION_TABLE for a mainboard
which depends on the CMOS backend for options.  This is needed to
support update mechanism based on Top Swap that permits recovery via a
CMOS reset (see [0]).  The indirection is necessary because
USE_OPTION_TABLE is part of a `choice`.

[0]: https://mail.coreboot.org/archives/list/coreboot@coreboot.org/thread/C6JN2PB7K7D67EG7OIKB6BBERZU5YV35/

Change-Id: I9e8f044077a5158650627a305c352cc9de578293
Upstream-Status: Backport [CB:90433]
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>
Reviewed-on: https://review.coreboot.org/c/coreboot/+/90433
Reviewed-by: Michał Kopeć <michal.kopec@3mdeb.com>
Tested-by: build bot (Jenkins) <no-reply@coreboot.org>
Reviewed-by: Filip Lewiński <filip.lewinski@3mdeb.com>
Reviewed-by: Filip Gołaś <filip.golas@3mdeb.com>

SergiiDmytruk committed:
cpu/intel/fit/Makefile.mk: make FIT in TOPSWAP point at MCU in COREBOOT_TS

This is a correction of CB:89570 (commit 04ea4724e2d0 ("Makefile.mk:
separate bootblocks into BOOTBLOCK and TOPSWAP")) which has FITs in both
BOOTBLOCK and TOPSWAP point at microcode in COREBOOT region.

This can be tested by comparing outputs of
    build/util/cbfstool/ifittool -f build/coreboot.rom -r TOPSWAP -D
and
    build/util/cbfstool/ifittool -f build/coreboot.rom -r BOOTBLOCK -D
with microcode addresses as shown by
    uefitool build/coreboot.rom
The addresses in two regions must not be identical and their last six
hex digits must match what uefitool shows in "Base:" field (not
"Offset:").

Change-Id: Ie37aee7a26be18d1a4d8993afd2a2484c38c0b1e
Upstream-Status: Backport [CB:90434]
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>
Reviewed-on: https://review.coreboot.org/c/coreboot/+/90434
Tested-by: build bot (Jenkins) <no-reply@coreboot.org>
Reviewed-by: Filip Gołaś <filip.golas@3mdeb.com>
Reviewed-by: Michał Kopeć <michal.kopec@3mdeb.com>
Reviewed-by: Filip Lewiński <filip.lewinski@3mdeb.com>

SergiiDmytruk committed:
Makefile.mk,cpu/intel/fit/Makefile.mk: introduce CBFS_REGIONS

Take advantage of cbfstool's ability to operate on multiple regions and
introduce a variable with a list of main CBFS regions: it's just
"COREBOOT" in most cases, but becomes "COREBOOT,COREBOOT_TS" when Top
Swap update mechanism is enabled (see [0]).

This is meant to simplify Makefiles by avoiding extra branches in
existing and future changes.

[0]: https://mail.coreboot.org/archives/list/coreboot@coreboot.org/thread/C6JN2PB7K7D67EG7OIKB6BBERZU5YV35/

Change-Id: If537d0d21a2867fafc2241ea9a0b4c0c6ca290a8
Upstream-Status: Backport [CB:90435]
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>
Reviewed-on: https://review.coreboot.org/c/coreboot/+/90435
Reviewed-by: Filip Lewiński <filip.lewinski@3mdeb.com>
Reviewed-by: Filip Gołaś <filip.golas@3mdeb.com>
Tested-by: build bot (Jenkins) <no-reply@coreboot.org>
Reviewed-by: Michał Kopeć <michal.kopec@3mdeb.com>

SergiiDmytruk committed:
Makefile.mk: don't add bootblock after other files

This is a correction of CB:89570 (commit 04ea4724e2d0 ("Makefile.mk:
separate bootblocks into BOOTBLOCK and TOPSWAP")).  It wasn't obvious
that CBFS verification depends on bootblock being added first (otherwise
cbfstool considers CBFS verification to be disabled because anchor is
part of a bootblock).  Correct this and add a comment for anyone else
who might edit this code in the future.

The issue manifested itself in build failing to perform CBFS
verification via cbfstool when the firmware was built with
CBFS_VERIFICATION enabled.

Change-Id: If775480394270fc05206cde0707c511b126265d3
Upstream-Status: Backport [CB:90436]
Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>
Reviewed-on: https://review.coreboot.org/c/coreboot/+/90436
Reviewed-by: Filip Lewiński <filip.lewinski@3mdeb.com>
Tested-by: build bot (Jenkins) <no-reply@coreboot.org>
Reviewed-by: Michał Kopeć <michal.kopec@3mdeb.com>

philipanda committed:
CBFS verification: support Top Swap redundancy

Separating the bootblock into two copies (in BOOTBLOCK and TOPSWAP fmap
regions) breaks the CBFS verification as TSPI CRTM knows nothing about
the new regions and looks for bootblock in a hard-coded COREBOOT fmap
region.

Introduce and use cbfs_unverified_area_type_alloc() which is an
extension of cbfs_unverified_area_alloc(), very similar to how
cbfs_ro_type_map() is an extension of cbfs_ro_map().  This allows to
specify a region of the bootblock file and skip verification because
bootblock serves as a container of hashes and is not verified itself.

The branching is done on the state of RTC BUC to always use the current
bootblock.  Somewhat confusingly, the measurement always uses BOOTBLOCK
region because with active Top Swap that's the way to access a
memory-mapped TOPSWAP region.

Makefile.mk now verifies both COREBOOT and COREBOOT_TS regions.
cbfstool needed a few updates as well:
 - recognize both "BOOTBLOCK" and "TOPSWAP" regions
 - recognize both "COREBOOT" and "COREBOOT_TS" regions
 - reset metadata cache before processing each region as cache may now
   be invalid

SMM doesn't link with vboot functions, so cbfs_file_hash_mismatch() has
to skip verification in SMM due to the use of CMOS options backend.

This is a part of the bootblock redundancy feature proposed
on the mailing list:
https://mail.coreboot.org/archives/list/coreboot@coreboot.org/thread/C6JN2PB7K7D67EG7OIKB6BBERZU5YV35/

Tested by successfully booting into Protectli VP6670 with Top Swap and
CBFS Verification features enabled and Top Swap state being toggled.

Change-Id: Ia75e714ae84d8c0ae09b27495e3056313b109999
Upstream-Status: Backport [CB:8961]
Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>
Reviewed-on: https://review.coreboot.org/c/coreboot/+/89691
Reviewed-by: Michał Kopeć <michal.kopec@3mdeb.com>
Reviewed-by: Filip Lewiński <filip.lewinski@3mdeb.com>
Tested-by: build bot (Jenkins) <no-reply@coreboot.org>

SergiiDmytruk commented on src/mainboard/protectli/vault_adl_p/top_swap.fmd:
This file can be simplified by removing vboot-related files (`WP_RO`, `RW_SECTION_A`, `RO_FRID*`, `NVRAM`, `VPD`).

filipleple replied:
[Simplified](https://github.com/Dasharo/coreboot/pull/812/changes/b4b1c3b60a727f19f604a9ef043ac4fe13bdfd93#diff-a722ee723e42acda369184b66a28ca43872599302b07c9914e7e3fb5a73275a7R22), while preserving 64KiB alignment of the crucial regions

SergiiDmytruk commented on src/mainboard/protectli/vault_adl_p/Kconfig:
```diff
@@ -68,7 +75,8 @@ config SOC_INTEL_CSE_SEND_EOP_EARLY
 config FMDFILE
-	default "src/mainboard/\$(CONFIG_MAINBOARD_DIR)/vboot-rwa.fmd" if VBOOT && VBOOT_SLOTS_RW_A
+	default "src/mainboard/\$(CONFIG_MAINBOARD_DIR)/vboot-rwa.fmd" if VBOOT
+	default "src/mainboard/\$(CONFIG_MAINBOARD_DIR)/top_swap.fmd" if INTEL_HAS_TOP_SWAP
```
`&& VBOOT_SLOTS_RW_A` part seems still relevant for `vboot-rwa.fmd`.  `INTEL_HAS_TOP_SWAP` is always set and should probably be changed to `!VBOOT`.

filipleple replied:
[Done](https://github.com/Dasharo/coreboot/pull/812/changes/b4b1c3b60a727f19f604a9ef043ac4fe13bdfd93#diff-a722ee723e42acda369184b66a28ca43872599302b07c9914e7e3fb5a73275a7R78)

SergiiDmytruk commented on src/mainboard/protectli/vault_adl_p/Kconfig:
```diff
@@ -17,6 +17,10 @@ config BOARD_SPECIFIC_OPTIONS
+	select HAVE_OPTION_TABLE
+	select INTEL_HAS_TOP_SWAP
+	select TOP_SWAP_REDUNDANCY
+	select INTEL_IFD_SET_TOP_SWAP_BOOTBLOCK_SIZE
```
```kconfig
select MAINBOARD_NEEDS_CMOS_OPTIONS if !VBOOT
```

filipleple replied:
`...NEEDS_CMOS...` is selected by `TOP_SWAP_REDUNDANCY` already:

```
config TOP_SWAP_REDUNDANCY
	bool "Toggle the Intel Top Swap - based redundancy"
	depends on HAVE_OPTION_TABLE && INTEL_HAS_TOP_SWAP
	select INTEL_ADD_TOP_SWAP_BOOTBLOCK
	select MAINBOARD_NEEDS_CMOS_OPTIONS
	select INTEL_TOP_SWAP_SEPARATE_REGIONS
	select INTEL_TOP_SWAP_OPTION_CONTROL

```

Though on the same note I think `select TOP_SWAP_REDUNDANCY` has to be guarded with `if !VBOOT`: https://github.com/Dasharo/coreboot/pull/812/changes/b4b1c3b60a727f19f604a9ef043ac4fe13bdfd93#diff-a722ee723e42acda369184b66a28ca43872599302b07c9914e7e3fb5a73275a7R22

filipleple committed:
src/mainboard/protectli/vault_adl_p/Kconfig: cleanup after redundancy

Fix the options to assure seamless builds for both VBOOT and Top Swap
configurations

Upstream-Status: Inappropriate [Dasharo downstream]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

SergiiDmytruk commented on src/lib/cbfs.c:
```diff
@@ -31,6 +31,11 @@ struct mem_pool cbfs_cache =
 	MEM_POOL_INIT(_cbfs_cache, REGION_SIZE(cbfs_cache), CONFIG_CBFS_CACHE_ALIGN);
 #endif
 
+__weak const char *cbfs_fmap_region_hint(const char *default_region)
```
Upstreamed version doesn't have the parameter.

filipleple replied:
Ah right, didn't backport the final tweak, [fixed](https://github.com/Dasharo/coreboot/pull/812/changes#diff-f37ee0acf30359fa8a2e8fdaba7fd8295498831683ab81c2c56a3a09aacd6fddR13)

SergiiDmytruk commented on configs/config.protectli_vp66xx:
```diff
@@ -17,6 +15,7 @@ CONFIG_SOFTWARE_CONNECTION_MANAGER=y
 CONFIG_IFDTOOL_DISABLE_ME=y
 CONFIG_HAVE_ME_BIN=y
 CONFIG_ME_REGION_ALLOW_CPU_READ_ACCESS=y
+CONFIG_INTEL_ME_DISABLED_HAP=y
```
Was this added as a convenience during testing or there was some other reason?

filipleple replied:
Yup, only for convenience, removed: https://github.com/Dasharo/coreboot/pull/812/changes/9ffa45a6e90b5dc9126630ace473099620141a9c

filipleple committed:
soc/intel/common/block: build rtc into postcar

Build the appropriate rtc.c file into postcar, with Top Swap logic
overriding the default behavior.

This resolves the issue with Top Swap Redundancy where the updated Slot
B boot flow would stall at loading ramstage, due to romstage using the
default boot device selection logic, incorrectly using Slot A instead.

Change-Id: I721a9444d52e93f1a18bbca826d5ca680541c14d
Upstream-Status: Pending
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

SergiiDmytruk commented on src/mainboard/protectli/vault_adl_p/cmos.default:
This file isn't used, right?

filipleple replied:
Right, removed the file

filipleple committed:
configs/config.protectli_vp66xx: unset ME HAP disable

Unset the flag, as it was only used to facilitate capsule update
testing.

Change-Id: I88352ab138843dc5e161fd1ef3a6068ef702c999
Upstream-Status: Inappropriate [Dasharo downstream]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
src/lib/cbfs.c: fix cbfs_fmap_region_hint() call

Fix the function call to match the new `(void)` prototype, after
removing redundant parameter logic

Change-Id: I301c5a0d13c4f13f123e98da08874f07a1f9b5ab
Upstream-Status: Inappropriate [Dasharo downstream]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

SergiiDmytruk commented:
CI failed due to outdated EDK revision, so I rebased https://github.com/Dasharo/edk2/pull/286 and bumped the revision here.

SergiiDmytruk commented on src/mainboard/protectli/vault_adl_p/Kconfig:
```diff
@@ -17,6 +17,13 @@ config BOARD_SPECIFIC_OPTIONS
+	select INTEL_IFD_SET_TOP_SWAP_BOOTBLOCK_SIZE
+
+config INTEL_TOP_SWAP_BOOTBLOCK_SIZE
+	default 0x80000
```
This copy of `INTEL_TOP_SWAP_BOOTBLOCK_SIZE` needs to be removed.  The one below is from upstream's backport.  It breaks CBnT change because that PR modifies the latter one.

filipleple replied:
https://github.com/Dasharo/coreboot/pull/812/changes/911da54a57d927ca684cfb0792f26acc4f3b1b38

filipleple committed:
mainboard/protectli/vault_adl_p/Kconfig: remove duplicate flag

A Kconfig option got duplicated when rebasing, breaking the result
binary

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: I4604b825dd2912c5adf9fc0188087d6273d7da02
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

SergiiDmytruk reviewed (approved):
EDK part has been merged, can merge this after a rebase (`dasharo` has moved).
